### ISSUE TITLE
hls4ml predict() input type and INT8 models

### ISSUE THREAD

**Issue by MateusFauri** [fastmachinelearning/hls4ml#1530]
## Prerequisites
Please make sure to check off these prerequisites before submitting a bug report.
- [ ] Test that the bug appears on the current version of the master branch. Make sure to include the commit hash of the commit you checked out.
- [x] Check that the issue hasn't already been reported, by checking the currently open issues.
- [x] If there are steps to reproduce the problem, make sure to write them down below.
- [ ] If relevant, please include the hls4ml project files, which were created directly before and/or after the bug.

## summary
Why does hls4ml's predict() function only accept float or double inputs instead of integer types?

My neural network is trained and quantized using 8-bit integer inputs (INT8). However, when using predict(), I need to convert the input image to float32, even though the deployed network expects an 8-bit input.

Is there a reason why predict() does not support integer inputs directly? Is converting an INT8 image to float32 before calling predict() the expected workflow, or is there a way to pass the quantized input directly?

### Steps to Reproduce
I trained a quantizade Keras CNN using a dataset with uint8 image data, which I converted to int8 before training/inference.

I then loaded the trained QKeras model and converted it to an hls4ml model using the Vitis backend

```
import hls4ml

co = {}
_add_supported_quantized_objects(co)

q_model = load_model(
    "models/results/t1/8bits/best/model.h5",
    custom_objects=co
)

config = hls4ml.utils.config_from_keras_model(
    q_model,
    granularity="name",
    backend="Vitis"
)

hls_model_test = hls4ml.converters.convert_from_keras_model(
    q_model,
    hls_config=config,
    backend="Vitis",
    output_dir="../hls4ml_prjs/hls4ml_prj_qkeras_part2",
    part="xc7z020clg400-1",
)

hls_model_test.compile()
```
The test dataset X_test has an integer data type. When I try to run inference directly:
```
y_hls_2 = hls_model_test.predict(
    np.ascontiguousarray(X_test)
)
```
I get an error because predict() does not accept the integer input type.

However, if I explicitly convert the input to float32:
```
y_hls_2 = hls_model_test.predict(
    np.ascontiguousarray(X_test.astype(np.float32))
)
```
the prediction works correctly.

### Expected behavior
I would expect `hls_model_test.predict()` to accept an integer input when the model input is configured/quantized to a integer quantization.

### Actual behavior
`predict()` requires the input to be converted to float or double before inference.

**Comment by vloncar** [comment]
We don't support this now. What's missing is the mapping between python and the generated C++. There's some subtlety to handle to support integer and integer-like types, as we default to fixed-precision now. Until this is addressed the workaround you use is the only way.

**Comment by MateusFauri** [comment]
But for this issue, I need to train my network using fixed-point precision rather than integer precision, right?

I was trying to configure my layers to use int<8> for the weights and biases, matching the quantized model I trained with QKeras. However, when I compiled the model and ran predictions with the hls4ml model, the accuracy dropped significantly compared to the QKeras quantization-aware trained model.

**Comment by vloncar** [comment]
Your example doesn't show what you wanted to do, how was the model defined etc, so it is not easy to reason about it. I can only give you some general information. Check what has been generated (the C++/HLS code, you should be able to quickly find what the input type is. While you're at it also check the other types match what you expect, they are in `defines.h`. This will point you in the direction of investigation. The `predict()` simply takes the pointer of the numpy array and passes that to the generated C wrapper function, which in turn casts to the defined input data type (`ap_(u)int` or `ap_fixed` in general case), invokes the top function, and casts back before returning the data to python, so I doubt the issue is in the wrapper. float32 is wide enough to support the cast. The issue could be in the model not picking up the input type and defaulting to `ap_fixed<16,6>`, or it could be the data types used for the computation within the network are not wide enough. A common cause of discrepancy between `hls_model.predict()` and `keras_model.predict()` is the data types used for accumulators and activation tensors. QKeras simulates quantization, the underlying data is always in float32, only clamped to the range of the quantizer after the multiplications. In hls4ml, everything is always in fixed precision, so models that are not quantized fully, or if the conversion didn't detect the types properly likely won't have the types wide enough to cover the whole range of intermediate results (of matrix-vector multiplications). There are tools in hls4ml to debug that, see the hls4ml-tutorials repo for introduction to these concepts.

**Comment by MateusFauri** [comment]
Sorry, I think I left out some important information in my previous message.

My goal is to create an 8-bit integer-quantized CNN. I trained it using QKeras, which resulted in this network, with an accuracy of 0.8745 (it is not perfect, but it is sufficient for testing purposes).

```

def build_quantizers(bits: int) -> Quantizers:

    if bits == 1:
        q = quantizers.binary(alpha=1)
        return Quantizers(weights=q, activation=q)

    return Quantizers(
        weights=quantizers.quantized_bits(bits, bits - 1, alpha=1),
        activation=quantizers.quantized_bits(bits, bits - 1, alpha=1),
    )


def build_t1(bits: int) -> Sequential:
    q = build_quantizers(bits)

    model = Sequential()
    model.add(Input(shape=INPUT_SHAPE, name="network_input"))

    model.add(
        QConv2D(
            filters=32,
            kernel_size=(3, 3),
            strides=(2, 2),
            padding="same",
            kernel_quantizer=q.weights,
            bias_quantizer=q.weights,
            name="conv2d_1",
        )
    )
    model.add(BatchNormalization(name="batch_normalization_1"))
    model.add(QActivation(q.activation, name="activation_1"))

    model.add(
        QConv2D(
            filters=64,
            kernel_size=(1, 1),
            padding="valid",
            kernel_quantizer=q.weights,
            bias_quantizer=q.weights,
            name="conv2d_2",
        )
    )
    model.add(BatchNormalization(name="batch_normalization_2"))
    model.add(QActivation(q.activation, name="activation_2"))

    model.add(Flatten(name="flatten"))

    model.add(
        QDense(
            128,
            kernel_quantizer=q.weights,
            bias_quantizer=q.weights,
            name="dense_1",
        )
    )
    model.add(BatchNormalization(name="batch_normalization_3"))
    model.add(QActivation(q.activation, name="activation_3"))

    model.add(
        QDense(
            NUM_CLASSES,
            kernel_quantizer=q.weights,
            bias_quantizer=q.weights,
            activation="softmax",
            name="dense_output",
        )
    )

    return model

``` 

The issue is that even when I create a config.yaml that forces the layers to use int<8> weights and biases, the network still uses fixed<8,8,TRN,WRAP,0> in some cases. I also tried using the default configuration with granularity='name', but that did not work either.
However, my goal is specifically to synthesize the quantized layers using integer precision, rather than floating-point or fixed-point precision.
But the bigger problem is that after setting my configurations and compiling the model, the accuracy drops to 0.0458. This seems very strange to me, since the original model has an accuracy of around 0.87.

OBS: The input network is images in int8. Like i said in the first question

My main question is whether the fact that integer precision is not yet supported for prediction, as I mentioned above, also affects the synthesis and could be the reason why the accuracy is so low. If not, I can open another forum post specifically for this issue.

Regarding the tutorials, I have already looked through them. However, I could not find any that cover explicit integer quantization; they only cover fixed-point precision, such as fixed<16,...>, which unfortunately does not help with my specific use case.

**Comment by vloncar** [comment]
Your model isn't fully quantized. For example, you're using BatchNormalization layers that work in full floating point precision in Keras, but will pick up whatever you specified in the conversion config. Depending on the case, it may not get folded away into the preceding convolution weights and thus survive. It can then behave very differently from the QKeras one. Also, you're not having any nonlinearity in the example you provided, the activations are linear quantizers, so something is not right on that side either.

There isn't really any difference between fixed point and integer handling. In fact what you define is equivalent to `fixed<8,7>`, so much of the tutorial is still relevant. Like I said, check to see if the types (inputs/outputs and intermediate values) of each defined layer in the generated top function matches your expectations with the Keras model. There's profiling tools to help with that as well.

Also, just looking at the number of your filters and the size of the last layer (`128*64*INPUT_SIZE`), you're unlikely to be able to synthesize this model with hls4ml. Remember, all of it has to fit on the chip. Not possible on the chip you chose.

**Comment by kr1xh** [comment]
I would like to look into this issue

After reading the discussion, it seems to me that there could be two different issues: namely the Python `predict()` interface not accepting integer NumPy inputs, and the generated HLS model not using the requested integer precision consistently
I would request to look at these two cases first before going through the changes to the `predict()` path, by checking the `defines.h`/C++ types and comparing the model output with `float32` inputs to the QKeras reference. This can help ascertain if the issue is in the Python to C++ input mapping or in the generated layer/accumulator precision

If this separation makes sense, I am happy to look into a minimal test case and investigate the required dtype mapping

**Comment by MateusFauri** [comment]
Hi!

First, thank you @vloncar for your help. I initially tried to create a consistent quantized model in TensorFlow using 8-bit integer QAT, but I couldn't get good results. So, I decided to switch to PyTorch and Brevitas, using QONNX for the HLS4ML conversion.

The model is now working well in terms of quantization and accuracy. However, I'm still struggling to find the right HLS4ML configuration, as the accuracy drops significantly after compiling the model.

Just in case, here is the quantized CNN model I needed:

```
class CommonQuant(ExtendedInjector):
    bit_width_impl_type = BitWidthImplType.CONST
    scaling_impl_type = ScalingImplType.CONST
    restrict_scaling_type = RestrictValueType.FP
    zero_point_impl = ZeroZeroPoint
    float_to_int_impl_type = FloatToIntImplType.ROUND
    scaling_per_output_channel = False
    narrow_range = True
    signed = True

    @value
    def quant_type(bit_width):
        if bit_width is None:
            return QuantType.FP
        elif bit_width == 1:
            return QuantType.BINARY
        else:
            return QuantType.INT


class CommonWeightQuant(CommonQuant, WeightQuantSolver):
    scaling_const = 1.0


class CommonUintActQuant(Uint8ActPerTensorFloat):
    scaling_min_val = 2e-16
    bit_width = None
    restrict_scaling_type = RestrictValueType.LOG_FP

class quantizedCNN(nn.Module):
    def __init__(self, bit_quantization):
        super(t1_quantizedCNN, self).__init__()
        self.conv1 = qnn.QuantConv2d(
            4, 
            20, 
            kernel_size=3, stride=2, padding=1, 
            bias = False,
            weight_bit_width=bit_quantization, 
            weight_quant=CommonWeightQuant, 
        )
        self.relu1 = qnn.QuantReLU(
            act_quant=CommonUintActQuant,
            bit_width=bit_quantization,
            return_quant_tensor=True
        )
        
        self.conv2 = qnn.QuantConv2d(
            20, 
            8, 
            kernel_size=1, stride=1,
            bias = False,
            weight_bit_width=bit_quantization, 
            weight_quant=CommonWeightQuant, 
        )
        self.relu2 = qnn.QuantReLU(
            act_quant=CommonUintActQuant,
            bit_width=bit_quantization,
            return_quant_tensor=True
        )
        
        self.fc1 = qnn.QuantLinear(
            8*16*16, 
            6, 
            bias = False,
            weight_bit_width=bit_quantization, 
            weight_quant=CommonWeightQuant, 
        )

        if (bit_quantization == 2):
            for m in self.modules():
                if isinstance(m, qnn.QuantConv2d) or isinstance(m, qnn.QuantLinear):
                    nn.init.uniform_(m.weight.data, -1, 1)
                    if m.bias is not None:
                        nn.init.zeros_(m.bias.data)

    def forward(self, x):
        x = self.conv1(x)
        x = self.relu1(x)

        x = self.conv2(x)
        x = self.relu2(x)

        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        return x
```

And @kr1xh, there are actually two separate issues here. Sorry about that! 
At first, I thought they were correlated, but they actually weren't.

**Comment by kr1xh** [comment]
I would like to first reproduce the reported behavior and check the generated defines.h/C++ types and the Python-to-C++ mapping, and then identify what changes are needed to support int inputs and add regression tests. I would also like to check if the reported accuracy discrepancy is due to precision inference/conversion rather than the predict() input dtype itself.

Is that okay if I do this, @vloncar @MateusFauri?

**Comment by kr1xh** [comment]
Hiiiii @vloncar @MateusFauri , I investigated the integer-input side of this issue and was able to reproduce the problem locally.

I traced the failure to the Python → generated C++ boundary in `ModelGraph._get_top_function()` and the Vivado/Vitis bridge generation. Currently, `predict()` only maps NumPy `float32`/`float64` inputs to the generated wrapper, while integer model inputs require an explicit integer C++ input type.

I have a local implementation that:
- adds `np.int8` input handling in `_get_top_function()`;
- maps the Python `int8` input to `int8_t` at the bridge boundary;
- converts `int8_t` to the model's HLS input type;
- keeps the bridge output as `float`, preserving the existing `predict()` output behavior;
- adds regression tests covering both bridge generation and the NumPy to ctypes mapping.

The focused regression tests pass (`2 passed`).

The changes are available in my fork on branch `fix/int8-predict-input` (commit `9e2da190`).

I have not addressed the separate accuracy discrepancy mentioned in this issue.

I would appreciate feedback on whether this implementation approach is appropriate before I open a PR.
